object/saml.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 The casbin Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package object
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/astaxie/beego"
+	saml2 "github.com/russellhaering/gosaml2"
+	dsig "github.com/russellhaering/goxmldsig"
+)
+
+func ParseSamlResponse(samlResponse string) string {
+	certStore := dsig.MemoryX509CertificateStore{
+		Roots: []*x509.Certificate{},
+	}
+	samlResponse, _ = url.QueryUnescape(samlResponse)
+	de, err := base64.StdEncoding.DecodeString(samlResponse)
+	if err != nil {
+		panic(err)
+	}
+	deStr := strings.Replace(string(de), "\n", "", -1)
+	res := regexp.MustCompile(`<ds:X509Certificate>(.*?)</ds:X509Certificate>`).FindAllStringSubmatch(deStr, -1)
+	str := res[0][0]
```

Comment:
idp may pass in multiple public keys

Reply:
Yes, the metadata of Aliyun IDaaS is as below and you can find that the values of two `ds:X509Certificate` are the same. So `res[0][1]` and `res[0][0]` both work well.

```xml
<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="CASDOOR-TEST">
<md:IDPSSODescriptor WantAuthnRequestsSigned="false" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
<md:KeyDescriptor use="signing">
<ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
<ds:X509Data>
<ds:X509Certificate>MIIB5TCCAU6gAwIBAgIIRiyDuM0koEswDQYJKoZIhvcNAQEFBQAwNTELMAkGA1UEBhMCQ04xDzANBgNVBAgMBuWMl+S6rDEVMBMGA1UEAxMMQ0FTRE9PUi1URVNUMB4XDTIxMTEyOTEwMTIwMloXDTI0MTEyODEwMTIwMlowNTELMAkGA1UEBhMCQ04xDzANBgNVBAgMBuWMl+S6rDEVMBMGA1UEAxMMQ0FTRE9PUi1URVNUMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCKxhfTOmQWs7i5dudkFxOHktushcrUloteFAYsKFhmlqtmc6M6upUAr59Yw14f+6myql/pWRdJIHMgJQuEcWF01aj4uHmwRrFpJHT+AbMBxfW8VNGl6JopHTLirAaTDESLhNaYS1YWxPluHRJ+7k3OE02ixQG5GLlf6EH1gm/8/QIDAQABMA0GCSqGSIb3DQEBBQUAA4GBADg4loZ7nAkDJqVSQMY8R/IiXemDzbECssSQBtLQLySzM0zkYij07AOTHOVVD6qQs4kULK5zQGWWDywjLFtitnwRWZ/DE3X7dJ5Vo2SwsPNtMQ2GKKXpWsurzUBzA8zjaFjwA1aJEXY/0tNkQuSs/pyNjoNtIq61zdmMdicpwB8X</ds:X509Certificate>
</ds:X509Data>
</ds:KeyInfo>
</md:KeyDescriptor>
<md:KeyDescriptor use="encryption">
<ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
<ds:X509Data>
<ds:X509Certificate>MIIB5TCCAU6gAwIBAgIIRiyDuM0koEswDQYJKoZIhvcNAQEFBQAwNTELMAkGA1UEBhMCQ04xDzANBgNVBAgMBuWMl+S6rDEVMBMGA1UEAxMMQ0FTRE9PUi1URVNUMB4XDTIxMTEyOTEwMTIwMloXDTI0MTEyODEwMTIwMlowNTELMAkGA1UEBhMCQ04xDzANBgNVBAgMBuWMl+S6rDEVMBMGA1UEAxMMQ0FTRE9PUi1URVNUMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCKxhfTOmQWs7i5dudkFxOHktushcrUloteFAYsKFhmlqtmc6M6upUAr59Yw14f+6myql/pWRdJIHMgJQuEcWF01aj4uHmwRrFpJHT+AbMBxfW8VNGl6JopHTLirAaTDESLhNaYS1YWxPluHRJ+7k3OE02ixQG5GLlf6EH1gm/8/QIDAQABMA0GCSqGSIb3DQEBBQUAA4GBADg4loZ7nAkDJqVSQMY8R/IiXemDzbECssSQBtLQLySzM0zkYij07AOTHOVVD6qQs4kULK5zQGWWDywjLFtitnwRWZ/DE3X7dJ5Vo2SwsPNtMQ2GKKXpWsurzUBzA8zjaFjwA1aJEXY/0tNkQuSs/pyNjoNtIq61zdmMdicpwB8X</ds:X509Certificate>
</ds:X509Data>
</ds:KeyInfo>
</md:KeyDescriptor>
<md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://geonmahrnw.login.aliyunidaas.com/enduser/api/application/plugin_saml/idaas-cn-beijing-mu34ajc7ktlplugin_saml/sp_sso"/>
<md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://geonmahrnw.login.aliyunidaas.com/enduser/api/application/plugin_saml/idaas-cn-beijing-mu34ajc7ktlplugin_saml/sp_sso_post"/>
</md:IDPSSODescriptor>
</md:EntityDescriptor>
```